driver-core/src/main/java/com/datastax/driver/core/CodecRegistry.java

```diff
@@ -0,0 +1,760 @@
+        TypeCodec<T> codec = findOverridingCodec(cqlType, javaType, fqdn);
+        if (codec != null)
+            return codec;
+        return codecFor(cqlType, javaType);
+    }
+
+    /**
+     * Return an overriding {@link TypeCodec codec} that accepts the given {@link DataType CQL type}
+     * and the given value, and that applies only to the given keyspace, table and column.
+     * <p>
+     * If such an overriding codec cannot be found, the registry attempts to find
+     * a globally-registered one and returns it instead.
+     * <p>
+     * This method takes an actual Java object and tries to locate a suitable codec for it.
+     * For this reason, codecs must perform a {@link TypeCodec#accepts(Object) "manual" inspection}
+     * of the object to determine if they can accept it or not, which, depending on the implementations,
+     * can be an expensive operation; besides, the resulting codec cannot be cached.
+     * Therefore there might be a performance penalty when using this method.
+     * <p>
+     * Codecs returned by this method are NOT cached.
+     *
+     * @param cqlType The {@link DataType CQL type} the codec should accept; must not be {@code null}.
+     * @param value The value the codec should accept; must not be {@code null}.
+     * @param fqdn The fully-qualified name of the table column or UDT field to which the overriding codec applies;
+     * if {@code null}, no overriding codec will be located and the globally-registered codec will be returned instead.
+     * @return An overriding codec for the given keyspace, table and column, if any, otherwise, a globally-registered suitable codec.
+     * @throws CodecNotFoundException if an overriding codec is found but does not accept the given value;
+     * or if no suitable codec could be found at all.
+     */
+    public <T> TypeCodec<T> codecFor(DataType cqlType, T value, String fqdn) {
+        checkNotNull(cqlType, "Parameter cqlType cannot be null");
+        checkNotNull(value, "Parameter value cannot be null");
+        TypeCodec<T> codec = findOverridingCodec(cqlType, value, fqdn);
+        if (codec != null)
+            return codec;
+        return codecFor(cqlType, value);
+    }
+
+    private <T> TypeCodec<T> lookupCache(DataType cqlType, TypeToken<T> javaType) {
+        logger.debug("Looking up codec for {} <-> {}", cqlType, javaType);
+        CacheKey cacheKey = new CacheKey(javaType, cqlType);
+        try {
+            return (TypeCodec<T>)cache.get(cacheKey);
+        } catch (UncheckedExecutionException e) {
+            if(e.getCause() instanceof CodecNotFoundException)
+                throw (CodecNotFoundException)e.getCause();
+            throw new CodecNotFoundException(e.getCause(), cqlType, javaType);
+        } catch (ExecutionException e) {
+            throw new CodecNotFoundException(e.getCause(), cqlType, javaType);
+        }
+    }
+
+    private <T> TypeCodec<T> findCodec(DataType cqlType, TypeToken<T> javaType) {
```

Comment:
Ahh the differentiation between calling findCodec with a TypeToken or with a Value must explain my question earlier about how to differentiate UTF8 and Ascii codec.  Is that why there is this differentiation?